probe/sniff/sniffer.go

```diff
@@ -0,0 +1,247 @@
+			turnOn = time.After(off)        // enable the on switch
+			turnOff = nil                   // disable the off switch
+
+		case c := <-s.reports:
+			rpt.Sampling.Count = atomic.LoadUint64(&count)
+			rpt.Sampling.Total = atomic.LoadUint64(&total)
+			applySampling(rpt)
+			c <- rpt
+			atomic.StoreUint64(&count, 0)
+			atomic.StoreUint64(&total, 0)
+			rpt = report.MakeReport()
+
+		case <-done:
+			return
+		}
+	}
+}
+
+// applySampling compensates for sampling by artifically inflating counts
+// throughout the report. It should be run once, by the probe, before emitting
+// the report upstream.
+func applySampling(r report.Report) {
+	rate := r.Sampling.Rate()
+	if rate >= 1.0 {
+		return
+	}
+	factor := 1.0 / rate
+	for _, topology := range r.Topologies() {
+		for _, emd := range topology.EdgeMetadatas {
+			if emd.PacketCount != nil {
+				*emd.PacketCount = uint64(float64(*emd.PacketCount) * factor)
+			}
+			if emd.ByteCount != nil {
+				*emd.ByteCount = uint64(float64(*emd.ByteCount) * factor)
+			}
+		}
+	}
+}
+
+// Packet is an intermediate, decoded form of a packet, with the information
+// that the Scope data model cares about. Designed to decouple the packet data
+// source loop, which should be as fast as possible, and the process of
+// merging the packet information to a report, which may take some time and
+// allocations.
+type Packet struct {
+	SrcIP, DstIP       string
+	SrcPort, DstPort   string
+	Network, Transport int // byte counts
+}
+
+func (s *Sniffer) read(src gopacket.ZeroCopyPacketDataSource, dst chan Packet, process, total, count *uint64) {
+	var (
+		p    Packet
+		data []byte
+		err  error
+	)
+	for {
+		data, _, err = src.ZeroCopyReadPacketData()
+		if err == io.EOF {
+			return // done
+		}
+		if err != nil {
+			log.Printf("sniffer: read: %v", err)
+			continue
+		}
+		atomic.AddUint64(total, 1)
+		if atomic.LoadUint64(process) == 0 {
+			continue
+		}
+
+		if err := s.parser.DecodeLayers(data, &s.decoded); err != nil {
+			// We'll always get an error when we encounter a layer type for
+			// which we haven't configured a decoder.
+		}
+		for _, t := range s.decoded {
+			switch t {
+			case layers.LayerTypeEthernet:
+				//
+
+			case layers.LayerTypeICMPv4:
+				p.Network += len(s.icmp4.Payload)
+
+			case layers.LayerTypeICMPv6:
+				p.Network += len(s.icmp6.Payload)
+
+			case layers.LayerTypeIPv6:
+				p.SrcIP = s.ip6.SrcIP.String()
+				p.DstIP = s.ip6.DstIP.String()
+				p.Network += len(s.ip6.Payload)
+
+			case layers.LayerTypeIPv4:
+				p.SrcIP = s.ip4.SrcIP.String()
+				p.DstIP = s.ip4.DstIP.String()
+				p.Network += len(s.ip4.Payload)
+
+			case layers.LayerTypeTCP:
+				p.SrcPort = strconv.Itoa(int(s.tcp.SrcPort))
+				p.DstPort = strconv.Itoa(int(s.tcp.DstPort))
+				p.Transport += len(s.tcp.Payload)
+
+			case layers.LayerTypeUDP:
+				p.SrcPort = strconv.Itoa(int(s.udp.SrcPort))
+				p.DstPort = strconv.Itoa(int(s.udp.DstPort))
+				p.Transport += len(s.udp.Payload)
```

Comment:
For performance you might want to consider not converting ips & ports to strings until you have to produce the report - ie keep them as []byte and ints, merger based on that, and convert at the last step.